core/src/main/java/kafka/server/share/DelayedShareFetch.java

```diff
@@ -89,25 +91,45 @@ public void onComplete() {
+        if (topicPartitionDataFromTryComplete.isEmpty())
             topicPartitionData = acquirablePartitions();
-            if (topicPartitionData.isEmpty()) {
-                // No locks for share partitions could be acquired, so we complete the request with an empty response.
-                shareFetchData.future().complete(Collections.emptyMap());
-                return;
-            }
-            fetchResponseData = replicaManagerFetchData(topicPartitionData, true);
-        } else {
-            // tryComplete invoked forceComplete, so we can use the topic partitions data and replica manager response data from tryComplete.
+        // tryComplete invoked forceComplete, so we can use the data from tryComplete.
+        else
             topicPartitionData = topicPartitionDataFromTryComplete;
-            fetchResponseData = replicaManagerFetchDataFromTryComplete;
+
+        if (topicPartitionData.isEmpty()) {
+            // No locks for share partitions could be acquired, so we complete the request with an empty response.
+            shareFetchData.future().complete(Collections.emptyMap());
+            return;
         }
+        log.trace("Fetchable share partitions data: {} with groupId: {} fetch params: {}",
+                topicPartitionData, shareFetchData.groupId(), shareFetchData.fetchParams());
 
         try {
+            Seq<Tuple2<TopicIdPartition, LogReadResult>> responseLogResult = replicaManager.readFromLog(
+                shareFetchData.fetchParams(),
+                CollectionConverters.asScala(
+                    topicPartitionData.entrySet().stream().map(entry ->
+                        new Tuple2<>(entry.getKey(), entry.getValue())).collect(Collectors.toList())
+                ),
+                QuotaFactory.UnboundedQuota$.MODULE$,
+                true);
+
+            Map<TopicIdPartition, FetchPartitionData> responseData = new HashMap<>();
+            Map<TopicIdPartition, LogOffsetMetadata> fetchOffsetMetadata = new HashMap<>();
+            responseLogResult.foreach(tpLogResult -> {
+                TopicIdPartition topicIdPartition = tpLogResult._1();
+                LogReadResult logResult = tpLogResult._2();
+                FetchPartitionData fetchPartitionData = logResult.toFetchPartitionData(false);
+                responseData.put(topicIdPartition, fetchPartitionData);
+                fetchOffsetMetadata.put(topicIdPartition, logResult.info().fetchOffsetMetadata);
+                return BoxedUnit.UNIT;
+            });
+
+            log.trace("Data successfully retrieved by replica manager: {}", responseData);
             Map<TopicIdPartition, ShareFetchResponseData.PartitionData> result =
-                ShareFetchUtils.processFetchResponse(shareFetchData, fetchResponseData, sharePartitionManager, replicaManager);
+                ShareFetchUtils.processFetchResponse(shareFetchData, responseData, fetchOffsetMetadata, sharePartitionManager, replicaManager);
```

Comment:
It's kind of late to set the offset metadata here since we will acquire the fetch records and move the next fetch offset. This means the cached offset metadata doesn't match the next fetch offset.

I was thinking of doing the following. In `tryComplete`, if the offset metadata doesn't exist, we call `replicaManager.readFromLog` to populate the offset metadata. We can then proceed with the minByte estimation. If SharePartition.endOffset moves, we invalidate the offset metadata.